### PR TITLE
curvefs tools space-status

### PR DESCRIPTION
**Pull request by Cyber-SiKu**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by Cyber-SiKu** [comment]
recheck

File: curvefs/src/tools/curvefs_tool.h [hunk 1]
```diff
@@ -90,42 +100,91 @@ template <class ChannelT, class ControllerT, class RequestT, class ResponseT,
...
+        service_stub_func_ = service_stub_func;
+        InitHostsAddr();
         return 0;
     }
 
+    virtual int Init() {
+        channel_ = std::make_shared<ChannelT>();
+        controller_ = std::make_shared<ControllerT>();
+        // request[0] could be re-assigned by subclass
+        requestQueue_.push(RequestT());
```
Comment (wu-hanqing) [inline curvefs/src/tools/curvefs_tool.h:129]: push an empty request, and set request fields by requestQueue_.front() is not straightforward

File: curvefs/src/tools/curvefs_tool.h [hunk 2]
```diff
@@ -174,20 +218,40 @@ class CurvefsToolRpc : public CurvefsTool {
...
+     * If necessary, you can override RunCommand in a subclass:
+     *      CurvefsToolRpc::RunCommand();
+     *      RemoveFailHostFromHostAddr();
+     * Add the fail host in AfterSendRequestToHost:
+     *      failHostsAddr_.push_back();
      */
-    virtual bool AfterSendRequestToService(const std::string& host) = 0;
+    void RemoveFailHostFromHostAddr() {
+        for (auto const& i : failHostsAddr_) {
+            auto pos = std::find(hostsAddr_.begin(), hostsAddr_.end(), i);
```
Comment (wu-hanqing) [inline curvefs/src/tools/curvefs_tool.h:248]: use std::remove_if and std::erase

File: curvefs/src/tools/umountfs/curvefs_umountfs_tool.cpp [hunk 3]
```diff
@@ -55,39 +56,40 @@ int UmountfsTool::RunCommand() {
...
 int UmountfsTool::Init() {
-    CurvefsToolRpc::Init();
+    int ret = CurvefsToolRpc::Init();
 
-    curve::common::SplitString(FLAGS_mdsAddr, ",", &hostsAddressStr_);
-
-    request_->set_fsname(FLAGS_fsname);
-    request_->set_mountpoint(FLAGS_mountpoint);
+    // adjust the unique element of the queue
+    requestQueue_.front().set_fsname(FLAGS_fsname);
```
Comment (wu-hanqing) [inline curvefs/src/tools/umountfs/curvefs_umountfs_tool.cpp:73]: it's not straightforward, just create a request and push it into request queue

File: curvefs/src/tools/status/curvefs_metaserver_status.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+#include "src/common/string_util.h"
+
+namespace curvefs {
+namespace tools {
+namespace status {
+
+class MetaserverStatusTool : public StatusBaseTool {
+ public:
+    MetaserverStatusTool()
+        : StatusBaseTool(std::string(kMdsStatusCmd), std::string(kProgrameName),
```
Comment (wu-hanqing) [inline curvefs/src/tools/status/curvefs_metaserver_status.h:39]: kMetaserverStatusCmd ?

File: curvefs/src/tools/query/curvefs_copyset_query.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+namespace query {
+
+class CopysetQueryTool
+    : public CurvefsToolRpc<brpc::Channel, brpc::Controller,
+                            curvefs::mds::topology::GetCopysetInfoRequest,
+                            curvefs::mds::topology::GetCopysetInfoResponse,
+                            curvefs::mds::topology::TopologyService_Stub> {
+ public:
+    explicit CopysetQueryTool(bool show = true)
+        : CurvefsToolRpc(std::string(kCopysetQueryCmd)) {
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_copyset_query.h:51]: > std::string is redundant.

fix

File: curvefs/src/tools/query/curvefs_copyset_query.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                tmp.set_poolid(i.poolid());
+                tmp.set_copysetid(i.copysetid());
+                tmp.set_allocated_peer(
+                    new curvefs::common::Peer(i.leaderpeer()));
+                *request.add_copysetsinfo() = std::move(tmp);
+            }
+            status::CopysetStatusTool copysetStatustool(false);
+            copysetStatustool.Init(request);
+            ret = copysetStatustool.RunCommand();
+            if (show_) {
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_copyset_query.cpp:96]: > missing handle ret == false ?

ret is used to return to tell the caller that the command line failed to execute, so there is no need to deal with it

File: curvefs/src/tools/query/curvefs_fsinfo_list.h [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+namespace tools {
+namespace query {
+class FsInfoListTool
+    : public CurvefsToolRpc<brpc::Channel, brpc::Controller,
+                            curvefs::mds::ListClusterFsInfoRequest,
+                            curvefs::mds::ListClusterFsInfoResponse,
+                            curvefs::mds::MdsService_Stub> {
+ public:
+    explicit FsInfoListTool(bool show = true)
+        : CurvefsToolRpc(std::string(kFsInfoListCmd)) {
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_fsinfo_list.h:46]: > ditto

rm

File: curvefs/src/tools/space/curvefs_space_base_tool.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+namespace tools {
+namespace space {
+
+std::string ByteToStringByMagnitude(uint64_t byte) {
+    // Convert byte KB to a appropriate magnitude
+    // like 1024KB to 1 MB
+    std::stringstream ss;
+    ss.setf(std::ios::fixed);
+    ss.precision(2);  // 2 decimal places
+    if (byte >= 1024 * 1024 * 1024) {
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/space/curvefs_space_base_tool.cpp:35]: fix

File: curvefs/src/tools/space/curvefs_space_base_tool.cpp [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        // TB
+        ss << double(byte) / double(1024 * 1024 * 1024) << " TB";
+    } else if (byte >= 1024 * 1024) {
+        // GB
+        ss << double(byte) / double(1024 * 1024) << " GB";
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/space/curvefs_space_base_tool.cpp:40]: > ditto

fix

File: curvefs/src/tools/space/curvefs_space_base_tool.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+#define CURVEFS_SRC_TOOLS_SPACE_CURVEFS_SPACE_BASE_TOOL_H_
+
+#include <sstream>
+#include <string>
+
+namespace curvefs {
+namespace tools {
+namespace space {
+
+std::string ByteToStringByMagnitude(uint64_t byte);
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/space/curvefs_space_base_tool.h:33]: > `ToReadableSizes` ?

fix

File: curvefs/src/tools/space/curvefs_metadata_usage_tool.h [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+ * @details
+ */
+class MatedataUsageTool
+    : public CurvefsToolRpc<brpc::Channel, brpc::Controller,
+                            curvefs::mds::StatMetadataUsageRequest,
+                            curvefs::mds::StatMetadataUsageResponse,
+                            curvefs::mds::MdsService_Stub> {
+ public:
+    MatedataUsageTool()
+        : CurvefsToolRpc(std::string(kMetedataUsageCmd),
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/space/curvefs_metadata_usage_tool.h:54]: > ditto, std::string is redundant

fix

File: curvefs/src/tools/space/curvefs_metadata_usage_tool.cpp [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+                  << std::endl
+                  << "left: " << ByteToStringByMagnitude(response_->left())
+                  << std::endl;
+    }
+
+    return ret;
+}
+
+int MatedataUsageTool::Init() {
+    CurvefsToolRpc::Init();
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/space/curvefs_metadata_usage_tool.cpp:63]: > init results are not checked ?

fix

File: curvefs/src/tools/status/curvefs_cluster_copyset_status.cpp [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    std::cout << " [-mdsAddr=" << FLAGS_mdsAddr << "]"
+              << " [-metaserverAddr=" << FLAGS_mdsAddr << "]"
+              << " [-detail=" << FLAGS_detail << "]";
+    std::cout << std::endl;
+}
+
+int ClusterCopysetStatusTool::Init() {
+    int ret = 0;
+    getFsId_ = std::make_shared<query::FsInfoListTool>(false);
+    getFsId_->Init();
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/status/curvefs_cluster_copyset_status.cpp:43]: > init result are not checked

fix

File: curvefs/src/tools/status/curvefs_cluster_copyset_status.cpp [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    for (auto const i : copysetIdSet) {
+        listcopysetInfoRequest.add_copysetid(i);
+    }
+    std::queue<decltype(listcopysetInfoRequest)> queueListCopysetInfo;
+    queueListCopysetInfo.push(listcopysetInfoRequest);
+    getCopysetinfo_->SetRequestQueue(queueListCopysetInfo);
+    getCopysetinfo_->RunCommand();
+    auto copysetInfoLsit = getCopysetinfo_->GetResponse()->copysetinfo();
+
+    // get copysetSatus
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/status/curvefs_cluster_copyset_status.cpp:88]: > typo: copysetStatus

fix

File: curvefs/src/tools/status/curvefs_status_base_tool.h [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    int Init() override;
+    virtual int ProcessMetrics();
+
+ protected:
+    std::vector<std::string> hostAddr_;
+    std::vector<std::string> standbyHost_;
+    std::vector<std::string> errorHost_;  // not leader,not standby
+    std::vector<std::string> offlineHost_;
+    std::vector<std::string> leaderHost_;
+    std::vector<std::string> onlineHost_;
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/status/curvefs_status_base_tool.h:65]: > add suffix `s` to those names, like, `hostAddrs_`

fix

File: curvefs/src/tools/status/curvefs_status_base_tool.cpp [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+            std::cerr << "]." << std::endl;
+        }
+    } else if (show_) {
+        std::cout << "leader " << hostType_ << ": " << leaderHost_[0]
+                  << std::endl;
+    }
+
+    // standby host
+    if (show_) {
+        std::cout << "standy " << hostType_ << ": [ ";
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/status/curvefs_status_base_tool.cpp:109]: > typo: standby ?

fix

File: curvefs/src/tools/topology/curvefs_build_topology_tool.cpp [hunk 17]
```diff
@@ -42,8 +42,8 @@ void UpdateFlagsFromConf(curve::common::Configuration* conf) {
         LOG(INFO) << "conf: " << FLAGS_mds_addr;
     }
 
-    if (GetCommandLineFlagInfo("rpcTimeOutMs", &info) && info.is_default) {
-        conf->GetUInt32Value("rpcTimeoutMs", &FLAGS_rpcTimeOutMs);
+    if (GetCommandLineFlagInfo("rpcTimeOut", &info) && info.is_default) {
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/topology/curvefs_build_topology_tool.cpp:45]: > why remote suffix `Ms` ?

it defined by sean

File: curvefs/src/tools/status/curvefs_status_base_tool.cpp [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+            std::cerr << "error " << hostType_ << ": [";
+            for (auto const& i : errorHost_) {
+                std::cerr << i << " ";
+            }
+            std::cout << "]." << std::endl;
+        }
+    }
+
+    // offline host
+    if (offlineHost_.size() > 0) {
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/status/curvefs_status_base_tool.cpp:129]: > `if (!offlineHost.empty())`

fix

File: curvefs/src/tools/curvefs_tool.h [hunk 19]
```diff
@@ -90,42 +100,91 @@ template <class ChannelT, class ControllerT, class RequestT, class ResponseT,
...
+
+        // add need update FlagInfos
+        AddUpdateFlags();
+        UpdateFlags();
+        InitHostsAddr();
+        return 0;
+    }
+
+    virtual void SetRequestQueue(const std::queue<RequestT>& requestQueue) {
+        requestQueue_ = std::move(requestQueue);
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool.h:141]: > std::move is unnecessary

fix

File: curvefs/src/tools/curvefs_tool_define.h [hunk 20]
```diff
@@ -26,48 +26,92 @@
...
+    "mds-status: show the status of mds\n"
+    "metaserver-status: show the status of metaserver\n"
+    "mds-status: show the status of etcd\n"
+    "copyset-query: query copyset by copysetId\n"
+    "partition-list: list partition in fsId"
     "You can specify the config path by -confPath to avoid typing too many "
     "options\n";  // NOLINT
 
+/* Status Host Type */
+constexpr char kHostTypeMds[] = "mds";
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_define.h:88]: > should be consistent, all const or constexpr

fix

File: curvefs/src/tools/curvefs_tool_define.cpp [hunk 21]
```diff
@@ -21,12 +21,20 @@
...
+DEFINE_string(fsId, "1,2,3", "fs id");
 DEFINE_string(mountpoint, "127.0.0.1:/mnt/curvefs-umount-test",
               "curvefs mount in local path");
+DEFINE_uint64(rpcTimeout, 3000, "millisecond for rpc timeout");
+DEFINE_uint64(rpcRetryTimes, 5, "rpc retry times");
+DEFINE_string(copysetId, "1,2,3", "copyset id");
+DEFINE_bool(detail, false, "show more info mation");
 
 // topology
 DEFINE_string(mds_addr, "127.0.0.1:6700",
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_define.cpp:40]: > duplicate with line 24

this usded for build-topology i will fix it in future.

File: curvefs/src/tools/curvefs_tool_main.cpp [hunk 22]
```diff
@@ -41,6 +41,7 @@ DECLARE_int32(health_check_interval);
 int main(int argc, char** argv) {
     google::SetUsageMessage(curvefs::tools::kHelpStr);
     google::ParseCommandLineFlags(&argc, &argv, true);
+    google::InitGoogleLogging(argv[0]);
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_main.cpp:44]: > no need to init glog

this usded for build-topology

File: curvefs/src/tools/curvefs_tool_define.h [hunk 23]
```diff
@@ -93,9 +138,40 @@ void SetFlagInfo(curve::common::Configuration* conf,
                  google::CommandLineFlagInfo* info, const std::string& key,
                  FlagInfoT* flag);
 
+/* update flags */
 extern std::function<void(curve::common::Configuration*,
                           google::CommandLineFlagInfo*)>
     SetMdsAddr;
+extern std::function<void(curve::common::Configuration*,
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_define.h:145]: > why not directly define them as a normal funciton?

fix

File: curvefs/src/tools/curvefs_tool_metric.h [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+#include <iostream>
+#include <string>
+
+#include "curvefs/src/tools/curvefs_tool_define.h"
+
+namespace curvefs {
+namespace tools {
+
+enum class MetricStatusCode {
+    // sucess
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_metric.h:37]: > typo: success

fix

File: curvefs/src/tools/curvefs_tool_metric.h [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+    kOK = 0,
+    // metric not found
+    kNotFound = -1,
+    // other error
+    kOtherErr = -2,
+};
+
+class MetricClient {
+ public:
+    virtual ~MetricClient() {}
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_metric.h:47]: > unnecessary

fix

File: curvefs/src/tools/curvefs_tool_metric.cpp [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+
+        return MetricStatusCode::kOK;
+    }
+
+    bool needRetry =
+        (cntl.Failed() && cntl.ErrorCode() != EHOSTDOWN &&
+         cntl.ErrorCode() != ETIMEDOUT && cntl.ErrorCode() != brpc::ELOGOFF &&
+         cntl.ErrorCode() != brpc::ERPCTIMEDOUT);
+    uint64_t retryTimes = 0;
+    while (needRetry && retryTimes < FLAGS_rpcRetryTimes) {
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_metric.cpp:59]: > put first rpc call in while loop

it copy from old will fix in future

File: curvefs/src/tools/curvefs_tool_define.cpp [hunk 27]
```diff
@@ -21,12 +21,20 @@
...
+              "metaserver addr");
+DEFINE_string(etcdAddr, "127.0.0.1:2379", "etcd addr");
 DEFINE_bool(example, false, "print the example of usage");
-DEFINE_string(confPath, "curvefs/conf/tools.conf", "config file path of tools");
+DEFINE_string(confPath, "/etc/curvefs/tools.conf", "config file path of tools");
 DEFINE_string(fsname, "curvefs", "fs name");
+DEFINE_string(fsId, "1,2,3", "fs id");
 DEFINE_string(mountpoint, "127.0.0.1:/mnt/curvefs-umount-test",
               "curvefs mount in local path");
+DEFINE_uint64(rpcTimeout, 3000, "millisecond for rpc timeout");
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool_define.cpp:34]: > duplicate with line 43

this usded for build-topology fix in future

File: curvefs/src/tools/curvefs_tool.h [hunk 28]
```diff
@@ -90,42 +100,91 @@ template <class ChannelT, class ControllerT, class RequestT, class ResponseT,
...
+        : CurvefsTool(command, programe, show) {}
 
     int Init(const std::shared_ptr<ChannelT>& channel,
              const std::shared_ptr<ControllerT>& controller,
-             const std::shared_ptr<RequestT>& request,
+             const std::queue<RequestT>& requestQueue,
              const std::shared_ptr<ResponseT>& response,
-             const std::shared_ptr<ServiceT>& service_stub) {
+             const std::shared_ptr<ServiceT>& service_stub,
+             std::function<void(ControllerT*, RequestT*, ResponseT*)>
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool.h:113]: > const reference, treat std::function as std::string, it allocates resources

fix

File: curvefs/src/tools/curvefs_tool.h (hunk 2 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool.h:248]: > remove_if

fix

**Pull request closed without merging**